modules/internal/ChapelRange.chpl

```diff
@@ -689,41 +739,20 @@ module ChapelRange {
     return new range(i, b, true,  lw, hh, st, alt, ald);
   }
 
-
-  proc by(r : range(?i,?b,?s), step:chpl__unsignedType(i))
-  {
-    return chpl_by_help(r, step);
-  }
-
-  proc by(r : range(?i,?b,?s), step:chpl__signedType(i))
-  {
-    return chpl_by_help(r, step);
+  inline proc by(r, step) {
+    if !isRange(r) then
+      compilerError("the first argument of the 'by' operator is not a range");
+    return chpl_by(r, step, true);
   }
-
+  
   // We want to warn the user at compiler time if they had an invalid param
-  // stride rather than waiting until runtime. However, we don't want to stamp
-  // out a by function for every valid param stride, so these only handle
-  // invalid cases, and let the non-param cases above handle values known at
-  // compile time that are non-zero
-  proc by(r : range(?i,?b,?s), param step:chpl__unsignedType(i)) where step == 0
-  {
-    compilerError("the 'by' operator cannot take a value of zero");
-  }
-
-  proc by(r : range(?i,?b,?s), param step:chpl__signedType(i)) where step == 0
-  {
-    compilerError("the 'by' operator cannot take a value of zero");
-  }
-
-  proc by(r : range(?i,?b,?s), step)
-  {
-    compilerError("can't apply 'by' to a range with idxType ",
-                  typeToString(i), " using a step of type ",
-                  typeToString(step.type));
-    return r;
+  // stride rather than waiting until runtime.
+  inline proc by(r : range(?), param step) {
```

Comment:
Is there a reason for the two different signatures of "by" as in, why does one take a stride, and the other takes a generic r? Does it make more sense to have both of these versions take a range r, and then have a 3rd generic case that warns that the first argument isn't a range?